File: src/engines/numerology/services/calculator.py
"""
Specific logic for extracting Life Path, Expression, and Soul Urge numbers.
"""
from ..utils import parsers
from ..constants import parameters

def get_life_path(dob_str):
    """
    Calculates Life Path from YYYY-MM-DD string.
    1. Extract all digits.
    2. Sum them.
    3. Reduce.
    """
    # Extract only digits from "1992-07-23" -> "19920723"
    digits = [int(char) for char in dob_str if char.isdigit()]
    
    if not digits:
        raise ValueError("Date of birth must contain numbers.")

    initial_sum = sum(digits)
    return parsers.reduce_number(initial_sum)

def get_expression(full_name):
    """
    Calculates Expression Number from Full Name.
    1. Clean string (A-Z).
    2. Map letters to numbers.
    3. Sum and Reduce.
    """
    clean_name = parsers.clean_string(full_name)
    if not clean_name:
        raise ValueError("Name must contain letters.")
        
    numbers = [parameters.LETTER_MAP[char] for char in clean_name if char in parameters.LETTER_MAP]
    total = sum(numbers)
    return parsers.reduce_number(total)

def get_soul_urge(full_name):
    """
    Calculates Soul Urge Number from Vowels in Full Name.
    1. Clean string.
    2. Filter for Vowels (A, E, I, O, U).
    3. Map to numbers, Sum and Reduce.
    Returns None if no vowels found.
    """
    clean_name = parsers.clean_string(full_name)
    vowel_numbers = [
        parameters.LETTER_MAP[char] for char in clean_name 
        if char in parameters.VOWELS and char in parameters.LETTER_MAP
    ]
    
    if not vowel_numbers:
        return None
        
    total = sum(vowel_numbers)
    return parsers.reduce_number(total)

File: src/server/engines/numerology/calculator.ts
import { LETTER_MAP, MASTER_NUMBERS, VOWELS } from "./constants";

/** Strip non-alpha characters and uppercase. */
export function cleanString(s: string): string {
  return s.replace(/[^a-zA-Z]/g, "").toUpperCase();
}

/** Sum digits repeatedly until single digit or master number (11/22/33). */
export function reduceNumber(n: number): number {
  let value = Math.abs(n);
  while (value > 9 && !MASTER_NUMBERS.has(value)) {
    let sum = 0;
    while (value > 0) {
      sum += value % 10;
      value = Math.floor(value / 10);
    }
    value = sum;
  }
  return value;
}

/** Life path number from date of birth (YYYY-MM-DD). */
export function getLifePath(dob: string): number {
  const digits = dob.replace(/\D/g, "");
  if (digits.length === 0) {
    throw new Error("Invalid date of birth: no digits found");
  }
  const digitSum = digits.split("").reduce((sum, d) => sum + Number(d), 0);
  return reduceNumber(digitSum);
}

/** Expression number from full name (all letters mapped and summed). */
export function getExpression(fullName: string): number {
  const cleaned = cleanString(fullName);
  if (cleaned.length === 0) {
    throw new Error("Invalid name: no alphabetic characters found");
  }
  const letterSum = cleaned
    .split("")
    .reduce((sum, ch) => sum + (LETTER_MAP[ch] ?? 0), 0);
  return reduceNumber(letterSum);
}

/** Soul urge number from vowels in full name. Returns null if no vowels. */
export function getSoulUrge(fullName: string): number | null {
  const cleaned = cleanString(fullName);
  const vowelLetters = cleaned.split("").filter((ch) => VOWELS.has(ch));
  if (vowelLetters.length === 0) {
    return null;
  }
  const vowelSum = vowelLetters.reduce(
    (sum, ch) => sum + (LETTER_MAP[ch] ?? 0),
    0,
  );
  return reduceNumber(vowelSum);
}
